### PR TITLE
Create modeler api tutorial

### PR DESCRIPTION
**Pull request by christinaausley**
Drafting up Modeler API tutorial. Docs to follow once this is polished up, but a few questions.

### CODE REVIEW

File: .env.example
Comment (christinaausley) [inline .env.example]: Have I mixed up audience and base URL here?

File: modeler.js
Comment (christinaausley) [inline modeler.js]: After reading the Modeler API docs, it seems like all I need is the a shell script to obtain a token, but with the auth we have set up, am I correct that all we need is the client ID, client secret, and what I've outlined as the audience?

File: modeler.js
Comment (christinaausley) [inline modeler.js]: I think I mentioned this last week, but I'm just wondering how to format the output properly, and what we'd want to include.

**Comment by christinaausley** [comment]
@pepopowitz I think this is ready for a first glance, but I have a few loose ends I'm not sure about.

If you'd rather, feel free to pop a half hour on my calendar and we can discuss.

File: .env.example
Comment (pepopowitz) [reply .env.example]: Yes! The audience is probably `modeler.camunda.io`, but we can confirm that from the API docs.

File: modeler.js
Comment (pepopowitz) [reply modeler.js]: Yes, that's correct. I'm not sure what the shell script is doing to obtain a token, but from the perspective of the CLI that we're building, passing the id/secret/audience to the auth endpoint is the typical approach.

File: modeler.js
Comment (pepopowitz) [reply modeler.js]: I think this makes sense. We basically want enough information to identify the item. The user might want to display different things, but we leave that up to them, and we just demonstrate that we're getting critical information about the items.

File: modeler.js
Comment (pepopowitz) [reply modeler.js]: There are two issues with this line, I think: 

1. It's going to use the text `process.env.MODELER_AUDIENCE` verbatim, instead of injecting the value of that variable. Note the wrapping of that text in `${...}` in the below suggestion.
2. I'd expect it to be injecting the `MODELER_BASE_URL` variable, not the `MODELER_AUDIENCE`. 

I think the line should look like this, assuming the environment variables were properly configured: 

```suggestion
    const url = `${process.env.MODELER_BASE_URL}/files/${fileId}`;
```

**Comment by pepopowitz** [comment]
@christinaausley I'll give this demo a walkthrough locally today, to finalize a review!

**Comment by pepopowitz** [comment]
Note that this tutorial will need to be updated based on https://github.com/camunda/camunda-api-tutorials/pull/7, as in https://github.com/camunda/camunda-docs/pull/3708/.

File: modeler.js
Comment (christinaausley) [reply modeler.js]: I don't know that {name} is gonna work here, because it is crossed out in my code editor?

**Comment by christinaausley** [comment]
@pepopowitz I've cleaned this up to mirror the updated auth -- let me know if you want to review this async, or jump on a call 👍

File: modeler.js
Comment (christinaausley) [inline modeler.js]: I wasn't sure, do we still need a cluster ID for this one?

Comment (pepopowitz) [reply modeler.js]: No, we don't need a cluster ID for any modeler API calls!

**Comment by pepopowitz** [comment]
@christinaausley I wasn't sure where to place this feedback relative to the changed code, so I'm making a general comment. 

I propose that we implement at least one API call different than we currently are. 

We're implementing two methods in this tutorial: one to retrieve details of a specific file, one to delete a specific file. The challenge for the user is that they'll need to go into the Modeler to find a file suitable for deletion, and capture only its ID from the URL. The URLs for files in Modeler don't make it easy to capture the ID -- for example, the ID for `https://modeler.cloud.camunda.io/process-applications/8f413a9a-ba96-4908-8c03-7887ebf161f7--basic/` is 8f413a9a-ba96-4908-8c03-7887ebf161f7 (note that the app appends the name of the file to the ID). This then becomes _our_ problem, of having to explain to them how to capture that ID from the URL. Opportunities for things to go wrong, things that aren't really the main focus of the tutorial. 

I have two proposals -- one more involved than the other. 

1. Instead of the "retrieve" method, we provide a "list" method, like we've done in the other API tutorials. This would allow them to list out _all_ IDs from their new CLI, and to capture one there more easily than in the Modeler URL. 
2. Instead of the "files" resource, we implement the "projects" resource. The "projects" resource enables us to _create_ a test item more simply than the "files" resource. [Check out the specs](https://modeler.cloud.camunda.io/swagger-ui/index.html#/Projects/createProject) -- a new project requires only a `name` property, compared to a file which requires a valid `projectId`, `folderId`, and `contents`. That would allow us to implement all the methods we did in the Administration API tutorial (list, add, view, delete) -- which would allow for the user to follow the tutorial without affecting existing files, which they may or may not feel comfortable deleting. 

What do you think? I'm open to other suggestions too, including iterative approaches like "let's do these methods and then come back and do one of the above suggestions." I'm also happy to contribute any changes or help if you are annoyed by me kicking this review a month down the road, and then suggesting such drastic changes.

**Comment by christinaausley** [comment]
> 1. Instead of the "retrieve" method, we provide a "list" method, like we've done in the other API tutorials. This would allow them to list out _all_ IDs from their new CLI, and to capture one there more easily than in the Modeler URL.

@pepopowitz This is an easy fix! I can probably do this today. But, do you have a preference on how we handle this? Could always open a backlog issue for iterations on these tutorials.

**Comment by christinaausley** [comment]
@pepopowitz I ended up using POST for the first API call, so the user creates a semi-meaningless file, the output reveals the name and ID, and then the user deletes it with this ID. WDYT?

**Comment by pepopowitz** [comment]
> @pepopowitz I ended up using POST for the first API call, so the user creates a semi-meaningless file, the output reveals the name and ID, and then the user deletes it with this ID. WDYT?

Did that work for you against the actual API? When I was playing with the API in Insomnia, I was unable to create a file that didn't have contents.

**Comment by pepopowitz** [comment]
> > 1. Instead of the "retrieve" method, we provide a "list" method, like we've done in the other API tutorials. This would allow them to list out _all_ IDs from their new CLI, and to capture one there more easily than in the Modeler URL.
> 
> @pepopowitz This is an easy fix! I can probably do this today. But, do you have a preference on how we handle this? Could always open a backlog issue for iterations on these tutorials.

I don't have a preference. In case you'd like to handle it with a follow-up, I'll approve this one as-is, once I confirm that posting a file without contents functions against the real API.

**Comment by pepopowitz** [comment]
> > @pepopowitz I ended up using POST for the first API call, so the user creates a semi-meaningless file, the output reveals the name and ID, and then the user deletes it with this ID. WDYT?
> 
> Did that work for you against the actual API? When I was playing with the API in Insomnia, I was unable to create a file that didn't have contents.

No, this API call does not work. 

See my previous comment about the File resource in the modeler API. Creating a file has several requirements, all of which will make it more work for us compared to a resource like Folder. Most of the requirements are captured [in the swagger description](https://modeler.cloud.camunda.io/swagger-ui/index.html#/Files/createFile), though not all of them (like there need to be valid file contents). 

If you'd like to treat the swapping of resources from Files to Folders iteratively, then I recommend reverting the changes to create the file.

**Comment by christinaausley** [comment]
> No, this API call does not work. See my previous comment about the File resource in the modeler API. 

AH! I understand now. I'll just adjust the calls to the Projects 👍

**Comment by christinaausley** [comment]
This is ready for review @pepopowitz, but if you don't get to it today, I can have @conceptualshark and/or @akeller have a look while you're away 😄 

Upon approval, I'll update the documentation. Thanks for your help!

File: modeler.js
Comment (pepopowitz) [inline modeler.js]: non-blocking conversation: 

One thing I learned while testing this, and it is documented in [the Modeler API's swagger UI](https://modeler.cloud.camunda.io/swagger-ui/index.html#/Projects/createProject) -- when creating a new project through the API, it has no collaborators, and therefore can't be viewed through the Modeler UI until you add collaborators via a different API endpoint. I discovered this, because I couldn't find my project in the UI after creating it 😅

I'm not sure what that means for this function. Do you think it's fine that the user won't be able to find their project in the UI after creating it? _I_ tried to find mine; I imagine others might do the same. Should we add an argument to this function for the collaborator and make a call to the second endpoint for them after creating the project? Are projects maybe not the panacea that I thought they were, and we should look for _another_ resource that can be created with no extra effort?

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
  const modelerApiUrl = process.env.MODELER_BASE_URL;
```

This name aligns with what is in the env.example, and what we've done on most other APIs.

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
      name: projectName
```

The API fields are specified in [the swagger UI](https://modeler.cloud.camunda.io/swagger-ui/index.html#/Projects/createProject).

File: modeler.js
Comment (pepopowitz) [inline modeler.js:7]:
```suggestion
      `Project added! Name: ${newProject.name}. ID: ${newProject.id}.`
```

Again, this is the correct name, as defined in the swagger.

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
  const modelerApiUrl = process.env.MODELER_BASE_URL;
```

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
    const data = response.data;

    // Emit the project details.
    console.log("Project:", data);
```

This is a more accurate name for this variable.

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
  const modelerApiUrl = process.env.MODELER_BASE_URL;
```

**Review by pepopowitz — DISMISSED** [review]
IGNORE this review, the suggestions somehow ended up on the incorrect lines.

**Comment by pepopowitz** [comment]
Oh man, I tried to make all my suggestions from within VS code, and it botched all the line numbers. I'm going to dismiss my previous review after re-submitting through the UI so that it's not suggesting misaligned changes 😭

File: modeler.js
Comment (pepopowitz) [inline modeler.js]: non-blocking conversation: 

One thing I learned while testing this, and it is documented in [the Modeler API's swagger UI](https://modeler.cloud.camunda.io/swagger-ui/index.html#/Projects/createProject) -- when creating a new project through the API, it has no collaborators, and therefore can't be viewed through the Modeler UI until you add collaborators via a different API endpoint. I discovered this, because I couldn't find my project in the UI after creating it 😅

I'm not sure what that means for this function. Do you think it's fine that the user won't be able to find their project in the UI after creating it? _I_ tried to find mine; I imagine others might do the same. Should we add an argument to this function for the collaborator and make a call to the second endpoint for them after creating the project? Are projects maybe not the panacea that I thought they were, and we should look for _another_ resource that can be created with no extra effort?

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
  const modelerApiUrl = process.env.MODELER_BASE_URL;
```

This name aligns with what is in the env.example, and what we've done on most other APIs.

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
      name: projectName
```

The API fields are specified in [the swagger UI](https://modeler.cloud.camunda.io/swagger-ui/index.html#/Projects/createProject).

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
      `Project added! Name: ${newProject.name}. ID: ${newProject.id}.`
```

Again, this is the correct name, as defined in the swagger.

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
  const modelerApiUrl = process.env.MODELER_BASE_URL;
```

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
    const data = response.data;

    // Emit the project details.
    console.log("Project:", data);
```

This is a more accurate name for this variable.

File: modeler.js
Comment (pepopowitz) [inline modeler.js]:
```suggestion
  const modelerApiUrl = process.env.MODELER_BASE_URL;
```

**Review by pepopowitz — CHANGES_REQUESTED** [review]
Nicely done, again! I left a few blocking requests re: variable names; anything non-blocking is specified as such.

Thanks for your patience with my requests to refine this!

File: modeler.js
Comment (christinaausley) [reply modeler.js]: > Do you think it's fine that the user won't be able to find their project in the UI after creating it? 

No, I agree here. Especially from a beginner's perspective, this could lead to a lot of confusion on their end.

> Should we add an argument to this function for the collaborator and make a call to the second endpoint for them after creating the project?

The simple fix here is to do this, and likely add an admonition explaining why the script looks like this. Does it solve the issue? Yes. Does it provide a clean tutorial? Not really, but it would mean we could merge this, and add an issue to the backlog to clean up. A "meh" option IMO.

> Are projects maybe not the panacea that I thought they were, and we should look for another resource that can be created with no extra effort?

Would it make sense to have the users create a project, then add a collaborator (PUT), then search for a collaborator (POST)?

I know @pepopowitz is out for a few weeks, so I'll loop in @akeller. This isn't urgent, so if you have time to take a look and provide feedback, I might be able to move forward with this and merge in a solution before the August alpha 👍

Comment (akeller) [reply modeler.js]: This is a point of friction, and we should encourage users to create a project and immediately add a collaborator so they can find it in the UI. I would also like you to take this feedback to the Modeler API team. I do not see an obvious reason why you would want to create a project with no collaborators, so could we _elect_ one _automagically_? That would be a more intuitive experience IMO. 

I would chain these exactly as you mentioned, @christinaausley. And absolutely be explicit that if they don't add a collaborator, they will not see this in the UI. The earlier they build that habit, the less likely they are to fall into that trap.

File: .env.example
Comment (akeller) [inline .env.example]: Please be consistent here and follow the same pattern set by the Optimize and Admin APIs.

```
# Web Modeler API tutorial
```

Then, for each value the user has to fill in, provide a comment about where this value comes from unless this is exactly how it appears on the UI. I got the following when creating my org level credentials with Modeler Scopes. 

```
export CAMUNDA_CONSOLE_CLIENT_ID='xxx'
export CAMUNDA_CONSOLE_CLIENT_SECRET='xxx'
export CAMUNDA_OAUTH_URL='https://login.cloud.camunda.io/oauth/token'
export CAMUNDA_CONSOLE_BASE_URL='https://api.cloud.camunda.io'
export CAMUNDA_CONSOLE_OAUTH_AUDIENCE='api.cloud.camunda.io'
```

Comment (christinaausley) [reply .env.example]: Adjusted in latest commit 👍

File: modeler.js
Comment (christinaausley) [inline modeler.js]: I had a tough time ensuring I got this data and the response data correct, but I think I'm close.

File: modeler.js
Comment (christinaausley) [reply modeler.js]: Made these changes -- will add a note on the collaborator discussion when this is approved and I can clean up [the docs PR](https://github.com/camunda/camunda-docs/pull/3910) 👍

File: modeler.js
Comment (akeller) [reply modeler.js]: Did you have problems because you can't test/run this?

> Adds a new collaborator to a project or modifies the permission level of an existing collaborator.

> Note: Only users that are part of the authorized organization (see GET /api/v1/info) and logged in to Web Modeler at least once can be added to a project.

After reading this in the docs, does it make sense to include it with the tutorial? While we could use this as a teachable moment and let the call fail, I'm concerned about the experience of an individual doing an evaluation. Could they easily create a test user to see this collaboration work in action?

Comment (christinaausley) [reply modeler.js]: Ah, I see.

Hmm...I'm not sure what API calls we could make for this tutorial for a beginner without these little road blocks.

The alternative is to add disclaimers/admonitions throughout the tutorial to be clear on this, or require a larger number of prerequisites.

I'll do some more digging, and sync with @pepopowitz at the retreat on options.

Comment (pepopowitz) [reply modeler.js]: After taking another pass through the Modeler API docs, I feel good about the requests you've put together here. 

The only thing I might add is a "deleteProject" at the end, so that we don't pollute their environment with test data. I don't feel strongly enough to push back on that in a review, but if it resonates and you think it's worthwhile, cool.

I'm now going to take a walk through the tutorial you've written!

Comment (pepopowitz) [reply modeler.js]: hmm, I guess you haven't updated the tutorial docs based on these changes, so I'll just see if I can make this code all run.

**Review by pepopowitz — CHANGES_REQUESTED** [review]
While playing with the code in this tutorial, I think I settled on a more ideal approach. I've written the code for it in https://github.com/camunda/camunda-api-tutorials/pull/9. 

Basically, it encapsulates the creation of a project and the addition of a collaborator into one action. It will take a little more explanation in the docs, to tell them why we're adding a second API call to this action and to walk them through the code.

But I think this aligns with what a user of our APIs would likely do in this situation. It doesn't appear that there's any endpoint in the Modeler API that can be used without the pre-existence of some other entity. In this case, the pre-existence required is the own user's account, which I think we can assume.

File: modeler.js
Comment (christinaausley) [reply modeler.js]: @pepopowitz Since we were a bit back and forth on the changes, I wanted to hold off on the docs changes until we were confident. Will update once this PR is finalized 👍

**Comment by christinaausley** [comment]
> But I think this aligns with what a user of our APIs would likely do in this situation. It doesn't appear that there's any endpoint in the Modeler API that can be used without the pre-existence of some other entity. In this case, the pre-existence required is the own user's account, which I think we can assume.

@pepopowitz Thank you for this! I'll update the documentation once this PR is finalized, and Slack you if I have any questions about how you've set it up.

File: .env.example
Comment (christinaausley) [inline .env.example]: Note to self to remove cloud from the base URL.

**Review by pepopowitz — APPROVED** [review]
💪 💪 💪 💪 💪 💪 💪

**Pull request merged**
